gigascience/gigadb-website
# A11y - Guide pages

# Pull request for issue: #1379 

This is a pull request for the following functionalities:

* Across the board accessibility improvements to guidelines pages (http://gigadb.gigasciencejournal.com:9170/site/guide and sibling pages)
  * Fix heading hierarchy
  * Improve accessibility of workflow diagram:
    * wrapped image in a `<figure>` element with a `<figcaption>`
    * added alt attribute
    * Added detailed text description of the diagram content
  * Avoid repeating h1 text as it is a bad SEO practice
  * Created dynamic component for guidelines secondary navigation, as it was a bit too overwhelming to repeat the same changes for this widget in 7 different places.
    * Note: I've never used Yii or php before for this kind of task, so I particularly welcome feedback on this part, in case I'm misusing some convetion or best practice, e.g. file location, naming, etc.
    * wrap secondary navigation in `<nav>` and remove `role="tabpanel"` attributes
    * replace link for more semantic button for dropdown button, and removed the `role="button"` attribute
  * Remove inline styles in updated code and replace them by css classes
  * Relocate embedded styles for tables to external stylesheet
  * Wrap table headings in `<thead>` element
  * Updated the label of some links for specificity
  * Changed the way screen readers see abbreviations used on tables (R for required, ^ for compulsory, etc). Other approaches could be possible, but I decided to hide all abbrevriations from screen readers and instead make accessible the full word in each case (but hidden from displays). For example: 

![image](https://github.com/gigascience/gigadb-website/assets/143437854/af27f935-4d66-4f13-9103-889c91e02b0b)

In the above case, the screen will show "R" but screen readers will omit "R" and will read "recommended".

It would also be possible to use aria attributes such as aria-describedby to connect abbrevriations to its description. I felt that in this case it would be easier to let screen readers ignore them.

## How to test?

* Review each of the sibling guidelines pages using the AXE chrome extension and verify that there are no errors within the page content and the heading hierarchy does not skip heading levels (you might notice a small number of alerts, but they are safe to ignore)
* Use the guides secondary navigation to navigate across pages and verify it works and has visible and distinct focus state
* Give the ok to the description of the diagram, as it is entirely new text (it's a one to one description of the diagram content)

## How have functionalities been implemented?

The content replacements for the guidelines siblings pages (hidden abbrevriations and inclusion of full text visible to screen readers) are as follows:

```html

<p>* - Requirements are listed as R= Recommended, O= Optional. Note ^ denotes absolutely mandatory fields.</p>

<p aria-hidden="true">* - Requirements are listed as R= Recommended, O= Optional. Note ^ denotes absolutely mandatory fields.</p>

Requirement <sup>*<sup>

Requirement <sup aria-hidden="true">*<sup>

<td>
    R
</td>

<td>
    <span aria-hidden="true">R</span><span class="sr-only">recommended</span>
</td>

<td>
    O
</td>

<td>
    <span aria-hidden="true">O</span><span class="sr-only">optional</span>
</td>

<sup>^</sup>

<sup aria-hidden="true">^</sup><span class="sr-only">, absolutely mandatory field</span>
```

## Any issues with implementation?

None

## Any changes to automated tests?

None

## Any changes to documentation?

None

## Any technical debt repayment?

None

## Any improvements to CI/CD pipeline?

None
